File: android/src/main/kotlin/com/alinmapay/payment/FlutterAlinmapayPaymentPlugin.kt
package com.alinmapay.payment

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.util.Log

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result


import com.alinmapay.android.PaymentSDK
import com.alinmapay.android.Interface.PaymentCallback
import com.alinmapay.android.Theme.SDKButtonStyle
import com.alinmapay.android.Theme.SDKTheme
import com.alinmapay.android.Theme.SDKMerchantBranding
import com.alinmapay.android.Theme.SDKInputStyle
import com.alinmapay.android.Theme.SDKTextStyle
import com.alinmapay.android.model.Request.PaymentRequest
import com.alinmapay.android.model.SDKConfiguration

class FlutterAlinmapayPaymentPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var flutterAssets: FlutterPlugin.FlutterAssets? = null

    override fun onAttachedToEngine(
        @NonNull binding: FlutterPlugin.FlutterPluginBinding
    ) {
        flutterAssets = binding.flutterAssets

        channel = MethodChannel(
            binding.binaryMessenger,
            "flutter_alinmapay_payment"
        )

        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> initializeSDK(call, result)
            "startPayment" -> startPayment(call, result)
            else -> result.notImplemented()
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun initializeSDK(call: MethodCall, result: Result) {
        try {
            val args = call.arguments as? Map<String, Any>
            if (args == null) {
                result.error("INVALID_ARGUMENTS", "Arguments missing", null)
                return
            }

            val themeMap = args["theme"] as? Map<String, Any>
            val theme = createTheme(themeMap)

            val configuration = SDKConfiguration.Builder()
                .setTerminalId(args["terminalId"] as? String ?: "")
                .setPassword(args["password"] as? String ?: "")
                .setMerchantKey(args["merchantKey"] as? String ?: "")
                .setBaseUrl(args["baseUrl"] as? String ?: "")
                .setTheme(theme)
                .build()

            Log.d("AlinmaPay", "========== SDK Configuration ==========")
            Log.d("AlinmaPay", "Terminal ID : ${configuration.terminalId}")
            Log.d("AlinmaPay", "Base URL    : ${configuration.baseUrl}")
            Log.d("AlinmaPay", "MerchantKey : ${configuration.merchantKey}")
            Log.d("AlinmaPay", "=======================================")

            PaymentSDK.initialize(configuration)
            result.success(true)

        } catch (e: Exception) {
            result.error("INIT_ERROR", e.localizedMessage ?: "SDK initialization failed", null)
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun createTheme(map: Map<String, Any>?): SDKTheme {
        val theme = SDKTheme()
        if (map == null) return theme
        val context = activity ?: return theme
        (map["primaryColor"] as? String)?.let { parseColorSafely(it)?.let { c -> theme.setPrimaryColor(c) } }
        (map["backgroundColor"] as? String)?.let { parseColorSafely(it)?.let { c -> theme.setBackgroundColor(c) } }

        val brandingMap = map["merchantBranding"] as? Map<String, Any>
        val branding = if (brandingMap != null) {
            when (brandingMap["type"] as? String) {
                "logo" -> {
                    val bitmap = getLogoBitmap(brandingMap["logo"] as? String)

                    if (bitmap != null) {
                        val drawable = BitmapDrawable(
                            context.resources,
                            bitmap
                        )

                        SDKMerchantBranding.logo(drawable)
                    } else {
                        SDKMerchantBranding.none()
                    }
                }
                "text" -> SDKMerchantBranding.text(brandingMap["text"] as? String ?: "")
                else -> SDKMerchantBranding.none()
            }
        } else {
            SDKMerchantBranding.none()
        }
        theme.setMerchantBranding(branding)

        val primaryButtonMap = map["primaryButton"] as? Map<String, Any>
        val primaryButtonStyle = SDKButtonStyle()
        primaryButtonMap?.let { btn ->
            (btn["backgroundColor"] as? String)?.let { parseColorSafely(it)?.let { c -> primaryButtonStyle.setBackgroundColor(c) } }
            (btn["textColor"] as? String)?.let { parseColorSafely(it)?.let { c -> primaryButtonStyle.setTextColor(c) } }
            (btn["borderColor"] as? String)?.let { parseColorSafely(it)?.let { c -> primaryButtonStyle.setBorderColor(c) } }
            (btn["borderWidth"] as? Number)?.let { primaryButtonStyle.setBorderWidth(it.toFloat()) }
            (btn["cornerRadius"] as? Number)?.let { primaryButtonStyle.setCornerRadius(it.toFloat()) }
            (btn["height"] as? Number)?.let { primaryButtonStyle.setHeight(it.toFloat()) }
        } ?: run {
            primaryButtonStyle.setCornerRadius(12f)
            primaryButtonStyle.setBorderWidth(0f)
            primaryButtonStyle.setHeight(50f)
        }
        theme.setPrimaryButtonStyle(primaryButtonStyle)

        val textStyleMap =
            map["sdkTextStyle"] as? Map<String, Any>

        val textColor =
            (textStyleMap?.get("textColor") as? String)
                ?.let { parseColorSafely(it) }
                ?: Color.TRANSPARENT

        val textSize =
            (textStyleMap?.get("textSize") as? Number)
                ?.toInt()
                ?: 26

        val bold =
            textStyleMap?.get("bold") as? Boolean
                ?: false

        val textStyle = SDKTextStyle(
            textSize,
            bold,
            textColor
        )

        theme.setTextStyle(textStyle)

        val inputStyleMap =
            map["inputStyle"] as? Map<String, Any>

        val inputBackgroundColor =
            (inputStyleMap?.get("backgroundColor") as? String)
                ?.let { parseColorSafely(it) }
                ?: Color.WHITE

        val inputBorderColor =
            (inputStyleMap?.get("borderColor") as? String)
                ?.let { parseColorSafely(it) }
                ?: Color.TRANSPARENT

        val inputBorderWidth =
            (inputStyleMap?.get("borderWidth") as? Number)
                ?.toFloat()
                ?: 1f

        val inputCornerRadius =
            (inputStyleMap?.get("cornerRadius") as? Number)
                ?.toFloat()
                ?: 12f

        val inputStyle = SDKInputStyle(
            inputBackgroundColor,
            inputBorderColor,
            inputBorderWidth,
            inputCornerRadius
        )

        theme.setInputStyle(inputStyle)

        return theme
    }

    private fun parseColorSafely(value: String): Int? {
        return try { Color.parseColor(value) } catch (e: Exception) { Log.e("AlinmaPay", "Invalid color: $value"); null }
    }

    private fun getLogoBitmap(logoName: String?): Bitmap? {
        if (logoName.isNullOrBlank()) {
            return null
        }

        val context = activity ?: return null
        val assets = flutterAssets ?: return null

        return try {
            val assetPath: String =
                assets.getAssetFilePathByName(logoName)

            Log.d(
                "AlinmaPay",
                "Loading Flutter logo: $logoName"
            )

            Log.d(
                "AlinmaPay",
                "Resolved asset path: $assetPath"
            )

            context.assets.open(assetPath).use { inputStream ->
                BitmapFactory.decodeStream(inputStream)
            }

        } catch (e: Exception) {
            Log.e(
                "AlinmaPay",
                "Failed to load logo: $logoName",
                e
            )
            null
        }
    }
    private fun startPayment(call: MethodCall, result: Result) {
        val currentActivity = activity ?: run { result.error("NO_ACTIVITY", "Activity is null", null); return }
        try {
            val request = PaymentRequest.Builder()
                .setAmount(call.argument<String>("amount") ?: "")
                .setTransactionType(call.argument<String>("transactionType") ?: "")
                .setCurrency(call.argument<String>("currency") ?: "")
                .setEmail(call.argument<String>("email") ?: "")
                .setAddress(call.argument<String>("address") ?: "")
                .setCity(call.argument<String>("city") ?: "")
                .setState(call.argument<String>("state") ?: "")
                .setZip(call.argument<String>("zip") ?: "")
                .setCountryCode(call.argument<String>("countryCode") ?: "")
                .setTrackId(call.argument<String>("trackId") ?: "")
                .setCardOperation(call.argument<String>("cardOperation") ?: "")
                .setCardToken(call.argument<String>("cardToken") ?: "")
                .setTokenType(call.argument<String>("tokenType") ?: "0")
                .setTransactionId(call.argument<String>("transactionId") ?: "")
                .setMetadata(call.argument<String>("metadata") ?: "")
                .build()

            PaymentSDK.getInstance().startPayment(currentActivity, request, object : PaymentCallback {
                override fun onResult(response: String) {
                    currentActivity.runOnUiThread { result.success(response) }
                }
            })
        } catch (e: Exception) {
            result.error("PAYMENT_ERROR", e.localizedMessage ?: "Payment failed", null)
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) { activity = binding.activity }
    override fun onDetachedFromActivity() { activity = null }
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) { activity = binding.activity }
    override fun onDetachedFromActivityForConfigChanges() { activity = null }
    override fun onDetachedFromEngine(
        @NonNull binding: FlutterPlugin.FlutterPluginBinding
    ) {
        channel.setMethodCallHandler(null)
        flutterAssets = null
    }
}

File: ios/Classes/FlutterAlinmapayPaymentPlugin.swift
import Flutter
import UIKit
import SwiftUI
import PaymentSDK

public class FlutterAlinmapayPaymentPlugin: NSObject, FlutterPlugin {

  private var flutterResult: FlutterResult?
  private var registrar: FlutterPluginRegistrar?
  public static func register(with registrar: FlutterPluginRegistrar) {

    let channel = FlutterMethodChannel(
      name: "flutter_alinmapay_payment",
      binaryMessenger: registrar.messenger()
    )

    let instance = FlutterAlinmapayPaymentPlugin()

    // Store registrar so Flutter assets can be resolved later
    instance.registrar = registrar

    registrar.addMethodCallDelegate(
      instance,
      channel: channel
    )
  }

  public func handle(
  _ call: FlutterMethodCall,
  result: @escaping FlutterResult
  ) {

    self.flutterResult = result

    switch call.method {

    case "initialize":
      initializeSDK(arguments: call.arguments)
      result(true)

    case "startPayment":
      startPayment(arguments: call.arguments, result: result)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Initialize SDK

  private func initializeSDK(arguments: Any?) {

    guard let args = arguments as? [String: Any] else {
      return
    }

    let themeMap = args["theme"] as? [String: Any]

    let theme = createTheme(from: themeMap)

    let config = SDKConfiguration(
      terminalId: args["terminalId"] as? String ?? "",
      password: args["password"] as? String ?? "",
      merchantKey: args["merchantKey"] as? String ?? "",
      baseURL: args["baseUrl"] as? String ?? "",
      theme: theme
    )

    print("========== SDK Configuration ==========")
    print("Terminal ID : \(config.terminalId)")
    print("Base URL    : \(config.baseURL)")
    print("MerchantKey : \(config.merchantKey)")
    print("=======================================")

    InAppSDK.shared.initialize(configuration: config)
  }

  // MARK: - Start Payment

  private func startPayment(
  arguments: Any?,
  result: @escaping FlutterResult
  ) {

    guard let args = arguments as? [String: Any] else {
      result(
        FlutterError(
          code: "INVALID_ARGUMENTS",
          message: "Arguments missing",
          details: nil
        )
      )
      return
    }

    let request = PaymentRequestData(
      amount: args["amount"] as? String ?? "",
      transactionType: args["transactionType"] as? String ?? "",
      currency: args["currency"] as? String ?? "",
      email: args["email"] as? String ?? "",
      address: args["address"] as? String ?? "",
      city: args["city"] as? String ?? "",
      state: args["state"] as? String ?? "",
      zip: args["zip"] as? String ?? "",
      countryCode: args["countryCode"] as? String ?? "",
      trackId: args["trackId"] as? String ?? "",
      cardOperation: args["cardOperation"] as? String ?? "",
      cardToken: args["cardToken"] as? String ?? "",
      tokenType: args["tokenType"] as? String ?? "",
      transactionId: args["transactionId"] as? String ?? "",
      metadata: args["metadata"] as? String ?? ""
    )

    InAppSDK.shared.startPayment(request: request) { response in

      DispatchQueue.main.async {
          result(response.amount ?? "") //Raw Response = response.rawjson
      }
    }
  }

  // MARK: - Theme
  private func createTheme(from map: [String: Any]?) -> SDKTheme {

    guard let map = map else {
      return .light
    }

    // MARK: Merchant Branding
    let branding: SDKMerchantBranding

    if let brandingMap = map["merchantBranding"] as? [String: Any] {

      let type = brandingMap["type"] as? String

      if type == "logo",
      let logoName = brandingMap["logo"] as? String {

        if let logo = getFlutterAssetImage(logoName) {
          branding = SDKMerchantBranding.logo(logo)
        } else {
          print("❌ Failed to load Flutter logo: \(logoName)")
          branding = SDKMerchantBranding.none()
        }

      } else if type == "text",
      let text = brandingMap["text"] as? String {

        branding = SDKMerchantBranding.text(text)

      } else {

        branding = SDKMerchantBranding.none()
      }

    } else {

      branding = SDKMerchantBranding.none()
    }

    // Primary Button

    let primaryButtonMap = map["primaryButton"] as? [String: Any]

//    let inputStyleModelMap = map["inputStyleModel"] as? [String: Any]
//    let sdkTextStyleModelMap = map["sdkTextStyleModel"] as? [String: Any]
//
//    // DEBUG
//    print("========== Flutter Theme ==========")
//    print("Theme Map: \(map)")
//    print("Primary Button Map: \(primaryButtonMap ?? [:])")
//    print("Primary Height: \(cgFloat(primaryButtonMap?["height"], defaultValue: 50))")
//
//    print("===================================")

    let primaryButtonStyle = SDKButtonStyle(
      backgroundColor: color(primaryButtonMap?["backgroundColor"]),
      textColor: color(primaryButtonMap?["textColor"]),
      cornerRadius: CGFloat(primaryButtonMap?["cornerRadius"] as? Double ?? 12),
      borderColor: color(primaryButtonMap?["borderColor"]),
      borderWidth: CGFloat(primaryButtonMap?["borderWidth"] as? Double ?? 0),

      height: cgFloat(primaryButtonMap?["height"], defaultValue: 50.0)
    )

    let textStyleMap =
    map["sdkTextStyle"] as? [String: Any]
    let textStyle =
    SDKTextStyle(

      color:
      color(
        textStyleMap?["textColor"]
      ),

      height:
      cgFloat(
        textStyleMap?["textSize"],
        defaultValue: 26
      ),

      isBold:
      textStyleMap?["bold"] as? Bool ?? false
    )
    // MARK: Input Style

    let inputStyleMap =    map["inputStyle"] as? [String: Any]

    let inputStyle =
            SDKInputStyle(
      backgroundColor: color(inputStyleMap?["backgroundColor"]),
      borderColor:
      color(
        inputStyleMap?["borderColor"]
      ),

      borderWidth:
      cgFloat(
        inputStyleMap?["borderWidth"],
        defaultValue: 1
      ),

      cornerRadius:
      cgFloat(
        inputStyleMap?["cornerRadius"],
        defaultValue: 12
      )
    )





    return SDKTheme (
      primaryColor: color(map["primaryColor"]),
      merchantBranding: branding,
      backgroundColor: color(map["backgroundColor"]),
      primaryButtonStyle: primaryButtonStyle,
      labelStyle: textStyle,
      inputStyle: inputStyle
    )
  }

  // MARK: - Flutter Asset Loader

  private func getFlutterAssetImage(_ assetName: String) -> UIImage? {

    guard let registrar = registrar else {
      print("❌ Flutter registrar is not available")
      return nil
    }

    let key = registrar.lookupKey(forAsset: assetName)

    print("========== Flutter Asset ==========")
    print("Asset Name : \(assetName)")
    print("Asset Key  : \(key)")
    print("===================================")

    guard let image = UIImage(named: key) else {
      print("❌ UIImage could not load asset: \(key)")
      return nil
    }

    print("✅ Flutter asset loaded successfully")

    return image
  }
  private func color(_ value: Any?) -> Color {

    guard let hex = value as? String else {
      return .clear
    }

    return Color(uiColor: UIColor(hex: hex))
  }


  private func cgFloat(_ value: Any?, defaultValue: CGFloat) -> CGFloat {

    if let value = value as? CGFloat {
      return value
    }

    if let value = value as? Double {
      return CGFloat(value)
    }

    if let value = value as? Float {
      return CGFloat(value)
    }

    if let value = value as? Int {
      return CGFloat(value)
    }

    if let value = value as? NSNumber {
      return CGFloat(truncating: value)
    }

    return defaultValue
  }
}

// MARK: - UIColor Extension

extension UIColor {

  convenience init(hex: String) {

    var hex = hex.trimmingCharacters(in: .whitespacesAndNewlines)

    hex = hex.replacingOccurrences(of: "#", with: "")

    var rgb: UInt64 = 0

    Scanner(string: hex).scanHexInt64(&rgb)

    self.init(
      red: CGFloat((rgb >> 16) & 0xFF) / 255.0,
      green: CGFloat((rgb >> 8) & 0xFF) / 255.0,
      blue: CGFloat(rgb & 0xFF) / 255.0,
      alpha: 1.0
    )
  }
}
